### PR TITLE
Do not promote to sharemem for dynamic tensors

### PR DESCRIPTION
**Pull request by okkwon**
Specialization helped the performance too and now the sharemem promotion actually slows down the partial tile portion. Needs a broad evaluation.

For example, distillbert shows improvement from 6.21ms to 5.29ms.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorAlloc.cpp [hunk 1]
```diff
@@ -22,6 +22,10 @@ namespace iree_compiler {
 static bool contractOpFilter(Operation *op) {
   auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
   if (!linalgOp) return false;
+
+  // Can't promote dynamic shapes.
+  if (linalgOp.hasDynamicShape()) return false;
```
[the remark this answers is not in the document]
Comment (okkwon) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorAlloc.cpp:29]: Thanks Thomas. Added more explanation.

**Comment by okkwon** [comment]
Got a perf result on A5000. Except for `layoutlm`, overall it improves the performance. 

| Benchmark | Before(ms) | After(ms) | Speedup |
|-----------|------------|-----------|---------|
|albert-base-v2 | 7.26 | 7.22 | 1.01 |
|bert-base-uncased | 17.30 | 16.20 | 1.07 |
|camembert-base | 17.70 | 16.70 | 1.06 |
|dbmdz/convert-base-turkish-cased | 14.60 | 14.70 | 0.99 |
|distilbert-base-uncased | 14.6 | 13.5 | 1.08 |
|gooble/electra-small-discriminator | 4 | 4 | 1 |
|gooble/mobilebert-uncased | 15.20 | 14.70 | 1.03 |
|microsoft/layoutlm-base-uncased | 14.60 | 16.9 | 0.86 |
|microsoft/MiniLM-L12-H384-uncased | 2.92 | 2.94 | 0.99 |
|resnet50 | 8.68 | 8.71 | 1 |
|roberta-base | 27.80 | 27.10 | 1.03|

**Comment by okkwon** [comment]
Created https://github.com/iree-org/iree/issues/10802 to further investigate the layoutlm regression.

**Review by ThomasRaoux — APPROVED** [review]
LGTM

**Comment by ThomasRaoux** [comment]
> Got a perf result on A5000. Except for `layoutlm`, overall it improves the performance.
> 
> Benchmark	Before(ms)	After(ms)	Speedup
> albert-base-v2	7.26	7.22	1.01
> bert-base-uncased	17.30	16.20	1.07
> camembert-base	17.70	16.70	1.06
> dbmdz/convert-base-turkish-cased	14.60	14.70	0.99
> distilbert-base-uncased	14.6	13.5	1.08
> gooble/electra-small-discriminator	4	4	1
> gooble/mobilebert-uncased	15.20	14.70	1.03
> microsoft/layoutlm-base-uncased	14.60	16.9	0.86
> microsoft/MiniLM-L12-H384-uncased	2.92	2.94	0.99
> resnet50	8.68	8.71	1
> roberta-base	27.80	27.10	1.03

nice, I wonder if the case that is slower is due to real dynamic shape that the specialization doesn't handle. That would get improved once you enable dynamic shapes.

**Pull request merged**